File: wrappers/javascript/src/exchange.js
'use strict';

/**
 * exchange.js — Exchange class for querying a single exchange calendar.
 *
 * This is the core class in the JavaScript wrapper. It represents one
 * exchange and answers questions like:
 *   - Is the market open right now?
 *   - Is this date a holiday?
 *   - Is this date an early close?
 *   - What time does the market close today?
 *   - What is the next trading day?
 *
 * The class is immutable after construction. All date/time arguments use
 * ISO 8601 date strings (YYYY-MM-DD) and 24-hour time strings (HH:MM).
 *
 * @example
 * const { Exchange } = require('./exchange');
 *
 * const xnys = new Exchange(exchangeData);
 * xnys.isOpen('2025-07-03', '10:00');   // true (before 13:00 early close)
 * xnys.isOpen('2025-07-03', '13:30');   // false (after early close)
 * xnys.isHoliday('2025-07-04');         // true (Independence Day)
 * xnys.earlyCloseTime('2025-07-03');    // "13:00"
 * xnys.nextTradingDay('2025-07-03');    // "2025-07-07" (Monday)
 */

const { SessionStatus } = require('./session');

class Exchange {
    /**
     * Create an Exchange from a registry data object.
     *
     * @param {Object} data — Exchange data as found in calendar.json.
     * @param {string} data.code — MIC code (e.g., "XNYS").
     * @param {string} data.name — Full exchange name.
     * @param {string} data.mic — ISO 10383 MIC, equal to code.
     * @param {string} data.timezone — IANA timezone (e.g., "America/New_York").
     * @param {Object} data.regular_hours — { open: "09:30", close: "16:00" }.
     * @param {Object} [data.extended_hours] — Pre-market and after-hours.
     * @param {Array} [data.sessions] — Auction and lunch break sessions.
     * @param {Object} data.holidays — { explicit: [...], generated: [...] }.
     *
     * @throws {Error} If required fields are missing or malformed.
     */
    constructor(data) {
        this._validateData(data);

        this.code = data.code;
        this.name = data.name;
        this.mic = data.mic;
        this.timezone = data.timezone;
        this.weekendDays = data.weekend_days || [5, 6];
        this.regularHours = data.regular_hours;
        this.extendedHours = data.extended_hours || {};
        this.sessions = data.sessions || [];

        const holidays = data.holidays || {};
        const explicit = holidays.explicit || [];
        const generated = holidays.generated || [];

        // Build lookup maps for O(1) date queries
        this._holidayByDate = new Map();
        this._statusByDate = new Map();
        this._earlyCloseTimeByDate = new Map();

        for (const entry of [...explicit, ...generated]) {
            this._indexEntry(entry);
        }
    }

    // ──────────────────────────────────────────────────────────
    // Data validation
    // ──────────────────────────────────────────────────────────

    _validateData(data) {
        if (!data || typeof data !== 'object') {
            throw new Error('Exchange: data must be an object');
        }

        const required = ['code', 'name', 'mic', 'timezone', 'regular_hours'];
        for (const field of required) {
            if (!(field in data)) {
                throw new Error(`Exchange: missing required field '${field}'`);
            }
        }

        if (!data.regular_hours.open || !data.regular_hours.close) {
            throw new Error("Exchange: regular_hours must have 'open' and 'close'");
        }

        Exchange._validateTimeFormat(data.regular_hours.open);
        Exchange._validateTimeFormat(data.regular_hours.close);

        if (data.code !== data.mic) {
            throw new Error(
                `Exchange: code '${data.code}' must equal mic '${data.mic}'`
            );
        }
    }

    static _validateTimeFormat(timeStr) {
        if (typeof timeStr !== 'string') {
            throw new Error(`Exchange: invalid time format '${timeStr}'. Expected HH:MM.`);
        }
        const match = timeStr.match(/^([01]\d|2[0-3]):([0-5]\d)$/);
        if (!match) {
            throw new Error(`Exchange: invalid time format '${timeStr}'. Expected HH:MM.`);
        }
    }

    static _validateDateFormat(dateStr) {
        if (typeof dateStr !== 'string') {
            throw new Error(`Exchange: invalid date format '${dateStr}'. Expected YYYY-MM-DD.`);
        }
        const match = dateStr.match(/^(\d{4})-(\d{2})-(\d{2})$/);
        if (!match) {
            throw new Error(`Exchange: invalid date format '${dateStr}'. Expected YYYY-MM-DD.`);
        }

        const year = parseInt(match[1], 10);
        const month = parseInt(match[2], 10);
        const day = parseInt(match[3], 10);

        const d = new Date(Date.UTC(year, month - 1, day));
        if (
            d.getUTCFullYear() !== year ||
            d.getUTCMonth() !== month - 1 ||
            d.getUTCDate() !== day
        ) {
            throw new Error(`Exchange: invalid date '${dateStr}'. Date does not exist.`);
        }
    }

    // ──────────────────────────────────────────────────────────
    // Internal indexing
    // ──────────────────────────────────────────────────────────

    _indexEntry(entry) {
        const dateStr = entry.date;
        const status = entry.status || 'closed';

        this._holidayByDate.set(dateStr, entry);
        this._statusByDate.set(dateStr, status);

        if (status === 'early_close' && entry.early_close_time) {
            this._earlyCloseTimeByDate.set(dateStr, entry.early_close_time);
        }
    }

    // ──────────────────────────────────────────────────────────
    // Date helpers
    // ──────────────────────────────────────────────────────────

    _isWeekend(dateStr) {
        Exchange._validateDateFormat(dateStr);
        const d = new Date(`${dateStr}T00:00:00Z`);
        // getUTCDay() is 0=Sunday..6=Saturday. weekend_days is stored using
        // 0=Monday..6=Sunday (matching the Python wrapper / stored data),
        // so convert before comparing — do not compare raw getUTCDay()
        // output against weekend_days directly.
        const isoDay = (d.getUTCDay() + 6) % 7; // 0=Monday..6=Sunday
        return this.weekendDays.includes(isoDay);
    }

    _isHoliday(dateStr) {
        Exchange._validateDateFormat(dateStr);
        return this._statusByDate.get(dateStr) === 'closed';
    }

    _isEarlyCloseDay(dateStr) {
        Exchange._validateDateFormat(dateStr);
        return this._earlyCloseTimeByDate.has(dateStr);
    }

    // ──────────────────────────────────────────────────────────
    // Public API — holiday queries
    // ──────────────────────────────────────────────────────────

    /**
     * Return true if the market is fully closed on this date.
     * Includes weekends and explicit/generated holidays.
     *
     * @param {string} dateStr — ISO date (YYYY-MM-DD).
     * @returns {boolean} True if market closed all day.
     */
    isHoliday(dateStr) {
        if (this._isWeekend(dateStr)) {
            return true;
        }
        return this._isHoliday(dateStr);
    }

    /**
     * Return true if this date has an early close.
     *
     * @param {string} dateStr — ISO date (YYYY-MM-DD).
     * @returns {boolean} True if market closes early.
     */
    isEarlyClose(dateStr) {
        return this._isEarlyCloseDay(dateStr);
    }

    /**
     * Return the early close time for this date, or null if not an early close.
     *
     * @param {string} dateStr — ISO date (YYYY-MM-DD).
     * @returns {string|null} Early close time as HH:MM, or null.
     */
    earlyCloseTime(dateStr) {
        return this._earlyCloseTimeByDate.get(dateStr) || null;
    }

    // ──────────────────────────────────────────────────────────
    // Public API — status
    // ──────────────────────────────────────────────────────────

    /**
     * Return the full session status at a specific date and time.
     *
     * Checks in order:
     *   1. Weekend → CLOSED
     *   2. Full holiday → CLOSED
     *   3. Early close day and time >= early_close_time → CLOSED
     *   4. Lunch break (if configured) → LUNCH_BREAK
     *   5. Before regular open → PRE_MARKET
     *   6. After regular close → AFTER_HOURS
     *   7. Otherwise → OPEN (or EARLY_CLOSE on early close day)
     *
     * @param {string} dateStr — ISO date (YYYY-MM-DD).
     * @param {string} timeStr — 24-hour time (HH:MM), interpreted as
     *   this exchange's LOCAL time (per its `timezone` field), NOT
     *   UTC and not the caller's local time. This wrapper does no
     *   timezone conversion -- `timezone` is exposed for
     *   informational purposes only and is not read by any
     *   status/date logic here. If you have a UTC or other-zone
     *   timestamp, convert it to this exchange's local time yourself
     *   before calling statusAt().
     * @returns {string} SessionStatus value.
     */
    statusAt(dateStr, timeStr) {
        Exchange._validateDateFormat(dateStr);
        Exchange._validateTimeFormat(timeStr);

        // 1. Weekend
        if (this._isWeekend(dateStr)) {
            return SessionStatus.CLOSED;
        }

        // 2. Full holiday
        if (this._isHoliday(dateStr)) {
            return SessionStatus.CLOSED;
        }

        // 3. Early close day — check if past the early close time
        const isEarlyCloseDay = this._isEarlyCloseDay(dateStr);
        if (isEarlyCloseDay) {
            const closeTime = this._earlyCloseTimeByDate.get(dateStr);
            if (timeStr >= closeTime) {
                return SessionStatus.CLOSED;
            }
        }

        // 4. Lunch break
        for (const session of this.sessions) {
            if (session.type === 'lunch_break') {
                const breakOpen = session.open;
                const breakClose = session.close;
                if (breakOpen && breakClose && breakOpen <= timeStr && timeStr < breakClose) {
                    return SessionStatus.LUNCH_BREAK;
                }
            }
        }

        // 5. Before regular open
        if (timeStr < this.regularHours.open) {
            return SessionStatus.PRE_MARKET;
        }

        // 6. After regular close
        if (timeStr >= this.regularHours.close) {
            return SessionStatus.AFTER_HOURS;
        }

        // 7. Within regular hours
        if (isEarlyCloseDay) {
            return SessionStatus.EARLY_CLOSE;
        }
        return SessionStatus.OPEN;
    }

    /**
     * Return true if the market is open for trading at the given moment.
     * Convenience wrapper around statusAt().
     *
     * @param {string} dateStr — ISO date (YYYY-MM-DD).
     * @param {string} [timeStr='10:00'] — 24-hour time (HH:MM).
     * @returns {boolean} True if trading is possible.
     */
    isOpen(dateStr, timeStr = '10:00') {
        const status = this.statusAt(dateStr, timeStr);
        return SessionStatus.isTradingStatus(status);
    }

    // ──────────────────────────────────────────────────────────
    // Public API — date navigation
    // ──────────────────────────────────────────────────────────

    /**
     * Return the next trading day after the given date.
     * Skips weekends and full holidays. Early close days count.
     *
     * @param {string} dateStr — ISO date (YYYY-MM-DD).
     * @returns {string} ISO date of the next trading day.
     */
    nextTradingDay(dateStr) {
        Exchange._validateDateFormat(dateStr);

        const d = new Date(`${dateStr}T00:00:00Z`);
        d.setUTCDate(d.getUTCDate() + 1);

        for (let i = 0; i < 30; i++) {
            const candidate = d.toISOString().slice(0, 10);
            if (!this.isHoliday(candidate)) {
                return candidate;
            }
            d.setUTCDate(d.getUTCDate() + 1);
        }

        throw new Error(`Exchange: no trading day found within 30 days after ${dateStr}`);
    }

    /**
     * Return the previous trading day before the given date.
     * Skips weekends and full holidays. Early close days count.
     *
     * @param {string} dateStr — ISO date (YYYY-MM-DD).
     * @returns {string} ISO date of the previous trading day.
     */
    previousTradingDay(dateStr) {
        Exchange._validateDateFormat(dateStr);

        const d = new Date(`${dateStr}T00:00:00Z`);
        d.setUTCDate(d.getUTCDate() - 1);

        for (let i = 0; i < 30; i++) {
            const candidate = d.toISOString().slice(0, 10);
            if (!this.isHoliday(candidate)) {
                return candidate;
            }
            d.setUTCDate(d.getUTCDate() - 1);
        }

        throw new Error(`Exchange: no trading day found within 30 days before ${dateStr}`);
    }

    // ──────────────────────────────────────────────────────────
    // Public API — metadata
    // ──────────────────────────────────────────────────────────

    /**
     * Return the number of explicit holidays in the registry.
     *
     * @param {number} [year] — Optional year filter.
     * @returns {number} Holiday count.
     */
    holidayCount(year = null) {
        if (year === null) {
            return this._holidayByDate.size;
        }

        const prefix = `${year}-`;
        let count = 0;
        for (const dateStr of this._holidayByDate.keys()) {
            if (dateStr.startsWith(prefix)) {
                count++;
            }
        }
        return count;
    }

    /**
     * Return a sorted list of holiday entries.
     *
     * @param {number} [year] — Optional year filter.
     * @returns {Array<Object>} Sorted holiday entries.
     */
    listHolidays(year = null) {
        const entries = Array.from(this._holidayByDate.values());

        if (year !== null) {
            const prefix = `${year}-`;
            return entries
                .filter(e => e.date.startsWith(prefix))
                .sort((a, b) => a.date.localeCompare(b.date));
        }

        return entries.sort((a, b) => a.date.localeCompare(b.date));
    }

    // ──────────────────────────────────────────────────────────
    // Dunder methods
    // ──────────────────────────────────────────────────────────

    toString() {
        return `${this.name} (${this.code})`;
    }

    toJSON() {
        return {
            code: this.code,
            name: this.name,
            mic: this.mic,
            timezone: this.timezone,
            regular_hours: this.regularHours,
            extended_hours: this.extendedHours,
            sessions: this.sessions,
            holiday_count: this._holidayByDate.size,
        };
    }
}

module.exports = { Exchange };

File: wrappers/go/exchange.go
package exchangecalendar

import (
	"fmt"
	"sort"
	"strings"
	"time"
)

// RegularHours represents the regular trading hours of an exchange.
type RegularHours struct {
	Open  string `json:"open"`  // HH:MM, e.g., "09:30"
	Close string `json:"close"` // HH:MM, e.g., "16:00"
}

// ExtendedHours represents pre-market and after-hours trading sessions.
type ExtendedHours struct {
	PreMarket  *RegularHours `json:"pre_market,omitempty"`
	AfterHours *RegularHours `json:"after_hours,omitempty"`
}

// Session represents a non-trading period within a regular trading day
// (e.g., lunch break or auction).
type Session struct {
	Type  string `json:"type"`            // "lunch_break", "auction", "other"
	Open  string `json:"open,omitempty"`  // Required for lunch_break
	Close string `json:"close,omitempty"` // Required for lunch_break
	At    string `json:"at,omitempty"`    // Required for auction
}

// HolidayEntry represents a single holiday or special session.
type HolidayEntry struct {
	Date           string `json:"date"`                      // YYYY-MM-DD
	Name           string `json:"name"`                      // Human-readable name
	Status         string `json:"status"`                    // "closed", "early_close", etc.
	EarlyCloseTime string `json:"early_close_time,omitempty"` // HH:MM when status is early_close
	DelayedOpenTime string `json:"delayed_open_time,omitempty"` // HH:MM when status is delayed_open
	SourceURL      string `json:"source_url,omitempty"`       // Source citation
}

// ExchangeData represents the raw JSON structure for one exchange in calendar.json.
type ExchangeData struct {
	Code          string         `json:"code"`
	Name          string         `json:"name"`
	MIC           string         `json:"mic"`
	Timezone      string         `json:"timezone"`
	WeekendDays   []int          `json:"weekend_days,omitempty"`
	RegularHours  RegularHours   `json:"regular_hours"`
	ExtendedHours ExtendedHours  `json:"extended_hours,omitempty"`
	Sessions      []Session      `json:"sessions,omitempty"`
	Holidays      HolidaysData   `json:"holidays"`
	AdHocClosures []HolidayEntry `json:"ad_hoc_closures,omitempty"`
	GenerationRange []string     `json:"generation_range,omitempty"`
}

// HolidaysData holds explicit and generated holiday entries.
type HolidaysData struct {
	Explicit  []HolidayEntry `json:"explicit"`
	Generated []HolidayEntry `json:"generated"`
}

// Exchange represents a single exchange calendar.
//
// The struct is immutable after construction. All exported fields are
// read-only by convention — do not modify them.
type Exchange struct {
	Code         string
	Name         string
	MIC          string
	Timezone     string
	WeekendDays  []int
	RegularHours RegularHours
	ExtendedHours ExtendedHours
	Sessions     []Session

	// Unexported lookup maps for O(1) queries
	holidayByDate       map[string]HolidayEntry
	statusByDate        map[string]string
	earlyCloseTimeByDate map[string]string
}

// NewExchange creates an Exchange from raw registry data.
//
// Returns an error if required fields are missing or malformed.
func NewExchange(data ExchangeData) (*Exchange, error) {
	if err := validateExchangeData(data); err != nil {
		return nil, err
	}

	e := &Exchange{
		Code:          data.Code,
		Name:          data.Name,
		MIC:           data.MIC,
		Timezone:      data.Timezone,
		WeekendDays:   weekendDaysOrDefault(data.WeekendDays),
		RegularHours:  data.RegularHours,
		ExtendedHours: data.ExtendedHours,
		Sessions:      data.Sessions,

		holidayByDate:        make(map[string]HolidayEntry),
		statusByDate:         make(map[string]string),
		earlyCloseTimeByDate: make(map[string]string),
	}

	// Index all holidays
	for _, entry := range append(data.Holidays.Explicit, data.Holidays.Generated...) {
		e.indexEntry(entry)
	}

	return e, nil
}

// MustNewExchange creates an Exchange or panics on error.
//
// Use only with known-valid data.
func MustNewExchange(data ExchangeData) *Exchange {
	e, err := NewExchange(data)
	if err != nil {
		panic(err)
	}
	return e
}

// ──────────────────────────────────────────────────────────────
// Internal validation
// ──────────────────────────────────────────────────────────────

func validateExchangeData(data ExchangeData) error {
	if data.Code == "" {
		return fmt.Errorf("exchange: missing code")
	}
	if data.Name == "" {
		return fmt.Errorf("exchange: missing name")
	}
	if data.MIC == "" {
		return fmt.Errorf("exchange: missing mic")
	}
	if data.Code != data.MIC {
		return fmt.Errorf("exchange: code %q must equal mic %q", data.Code, data.MIC)
	}
	if data.Timezone == "" {
		return fmt.Errorf("exchange: missing timezone")
	}
	if data.RegularHours.Open == "" || data.RegularHours.Close == "" {
		return fmt.Errorf("exchange: regular_hours must have open and close")
	}
	if err := validateTimeFormat(data.RegularHours.Open); err != nil {
		return fmt.Errorf("exchange: regular_hours.open: %w", err)
	}
	if err := validateTimeFormat(data.RegularHours.Close); err != nil {
		return fmt.Errorf("exchange: regular_hours.close: %w", err)
	}
	if data.RegularHours.Open >= data.RegularHours.Close {
		return fmt.Errorf("exchange: regular_hours.open (%s) must be before close (%s)",
			data.RegularHours.Open, data.RegularHours.Close)
	}
	return nil
}

func validateTimeFormat(timeStr string) error {
	if len(timeStr) != 5 {
		return fmt.Errorf("invalid time format %q: expected HH:MM", timeStr)
	}
	if timeStr[2] != ':' {
		return fmt.Errorf("invalid time format %q: expected HH:MM", timeStr)
	}
	hours := (timeStr[0]-'0')*10 + (timeStr[1] - '0')
	minutes := (timeStr[3]-'0')*10 + (timeStr[4] - '0')
	if hours > 23 || minutes > 59 {
		return fmt.Errorf("invalid time %q: hours must be 00-23, minutes 00-59", timeStr)
	}
	return nil
}

func validateDateFormat(dateStr string) error {
	if len(dateStr) != 10 {
		return fmt.Errorf("invalid date format %q: expected YYYY-MM-DD", dateStr)
	}
	_, err := time.Parse("2006-01-02", dateStr)
	if err != nil {
		return fmt.Errorf("invalid date %q: %w", dateStr, err)
	}
	return nil
}

// ──────────────────────────────────────────────────────────────
// Internal indexing
// ──────────────────────────────────────────────────────────────

func (e *Exchange) indexEntry(entry HolidayEntry) {
	e.holidayByDate[entry.Date] = entry
	e.statusByDate[entry.Date] = entry.Status

	if entry.Status == string(StatusEarlyClose) && entry.EarlyCloseTime != "" {
		e.earlyCloseTimeByDate[entry.Date] = entry.EarlyCloseTime
	}
}

// ──────────────────────────────────────────────────────────────
// Internal helpers
// ──────────────────────────────────────────────────────────────

// weekendDaysOrDefault falls back to Saturday/Sunday (in Monday=0..Sunday=6
// form) if the registry data omits weekend_days, for backward compatibility
// with older data files.
func weekendDaysOrDefault(wd []int) []int {
	if len(wd) == 0 {
		return []int{5, 6}
	}
	return wd
}

// isoWeekday converts Go's time.Weekday (Sunday=0..Saturday=6) into the
// Monday=0..Sunday=6 convention used by weekend_days in the registry data.
// Do not compare time.Weekday() directly against WeekendDays — the two
// use different day-numbering conventions.
func isoWeekday(d time.Time) int {
	return (int(d.Weekday()) + 6) % 7
}

func (e *Exchange) isWeekend(dateStr string) bool {
	d, err := time.Parse("2006-01-02", dateStr)
	if err != nil {
		return false // validateDateFormat should catch this before
	}
	day := isoWeekday(d)
	for _, wd := range e.WeekendDays {
		if day == wd {
			return true
		}
	}
	return false
}

// ──────────────────────────────────────────────────────────────
// Public API — holiday queries
// ──────────────────────────────────────────────────────────────

// IsHoliday returns true if the market is fully closed on this date.
// Includes weekends and explicit/generated holidays.
func (e *Exchange) IsHoliday(dateStr string) bool {
	if e.isWeekend(dateStr) {
		return true
	}
	return e.statusByDate[dateStr] == string(StatusClosed)
}

// IsEarlyClose returns true if this date has an early close.
func (e *Exchange) IsEarlyClose(dateStr string) bool {
	_, ok := e.earlyCloseTimeByDate[dateStr]
	return ok
}

// EarlyCloseTime returns the early close time for this date, or empty string.
func (e *Exchange) EarlyCloseTime(dateStr string) string {
	return e.earlyCloseTimeByDate[dateStr]
}

// ──────────────────────────────────────────────────────────────
// Public API — status
// ──────────────────────────────────────────────────────────────

// StatusAt returns the full session status at a specific date and time.
//
// Returns an error if date or time format is invalid.
//
// Checks in order:
//  1. Weekend → closed
//  2. Full holiday → closed
//  3. Early close day and time >= early_close_time → closed
//  4. Lunch break → lunch_break
//  5. Before regular open → pre_market
//  6. After regular close → after_hours
//  7. Otherwise → open (or early_close on early close day)
//
// timeStr (HH:MM) is interpreted as this exchange's LOCAL time (per
// its Timezone field), NOT UTC and not the caller's local time. This
// wrapper does no timezone conversion -- Timezone is exposed for
// informational purposes only and is not read by any status/date
// logic here. If you have a UTC or other-zone timestamp, convert it
// to this exchange's local time yourself before calling StatusAt.
func (e *Exchange) StatusAt(dateStr, timeStr string) (SessionStatus, error) {
	if err := validateDateFormat(dateStr); err != nil {
		return "", err
	}
	if err := validateTimeFormat(timeStr); err != nil {
		return "", err
	}

	// 1. Weekend
	if e.isWeekend(dateStr) {
		return StatusClosed, nil
	}

	// 2. Full holiday
	if e.statusByDate[dateStr] == string(StatusClosed) {
		return StatusClosed, nil
	}

	// 3. Early close day — check if past the early close time
	isEarlyCloseDay := e.IsEarlyClose(dateStr)
	if isEarlyCloseDay {
		closeTime := e.earlyCloseTimeByDate[dateStr]
		if timeStr >= closeTime {
			return StatusClosed, nil
		}
	}

	// 4. Lunch break
	for _, session := range e.Sessions {
		if session.Type == "lunch_break" && session.Open != "" && session.Close != "" {
			if session.Open <= timeStr && timeStr < session.Close {
				return StatusLunchBreak, nil
			}
		}
	}

	// 5. Before regular open
	if timeStr < e.RegularHours.Open {
		return StatusPreMarket, nil
	}

	// 6. After regular close
	if timeStr >= e.RegularHours.Close {
		return StatusAfterHours, nil
	}

	// 7. Within regular hours
	if isEarlyCloseDay {
		return StatusEarlyClose, nil
	}
	return StatusOpen, nil
}

// IsOpen returns true if the market is open for trading at the given moment.
// Convenience wrapper around StatusAt. Defaults timeStr to "10:00".
func (e *Exchange) IsOpen(dateStr string, timeStrs ...string) bool {
	timeStr := "10:00"
	if len(timeStrs) > 0 {
		timeStr = timeStrs[0]
	}
	status, err := e.StatusAt(dateStr, timeStr)
	if err != nil {
		return false
	}
	return status.IsTradingStatus()
}

// ──────────────────────────────────────────────────────────────
// Public API — date navigation
// ──────────────────────────────────────────────────────────────

// NextTradingDay returns the next trading day after the given date.
// Skips weekends and full holidays. Early close days count.
func (e *Exchange) NextTradingDay(dateStr string) (string, error) {
	if err := validateDateFormat(dateStr); err != nil {
		return "", err
	}

	d, _ := time.Parse("2006-01-02", dateStr)
	for i := 0; i < 30; i++ {
		d = d.AddDate(0, 0, 1)
		candidate := d.Format("2006-01-02")
		if !e.IsHoliday(candidate) {
			return candidate, nil
		}
	}
	return "", fmt.Errorf("no trading day found within 30 days after %s", dateStr)
}

// PreviousTradingDay returns the previous trading day before the given date.
// Skips weekends and full holidays. Early close days count.
func (e *Exchange) PreviousTradingDay(dateStr string) (string, error) {
	if err := validateDateFormat(dateStr); err != nil {
		return "", err
	}

	d, _ := time.Parse("2006-01-02", dateStr)
	for i := 0; i < 30; i++ {
		d = d.AddDate(0, 0, -1)
		candidate := d.Format("2006-01-02")
		if !e.IsHoliday(candidate) {
			return candidate, nil
		}
	}
	return "", fmt.Errorf("no trading day found within 30 days before %s", dateStr)
}

// ──────────────────────────────────────────────────────────────
// Public API — metadata
// ──────────────────────────────────────────────────────────────

// HolidayCount returns the number of holidays, optionally filtered by year.
func (e *Exchange) HolidayCount(year ...int) int {
	if len(year) == 0 {
		return len(e.holidayByDate)
	}

	prefix := fmt.Sprintf("%d-", year[0])
	count := 0
	for dateStr := range e.holidayByDate {
		if strings.HasPrefix(dateStr, prefix) {
			count++
		}
	}
	return count
}

// ListHolidays returns a sorted slice of holiday entries.
// Optionally filtered by year.
func (e *Exchange) ListHolidays(year ...int) []HolidayEntry {
	entries := make([]HolidayEntry, 0, len(e.holidayByDate))

	for _, entry := range e.holidayByDate {
		entries = append(entries, entry)
	}

	if len(year) > 0 {
		prefix := fmt.Sprintf("%d-", year[0])
		filtered := entries[:0]
		for _, entry := range entries {
			if strings.HasPrefix(entry.Date, prefix) {
				filtered = append(filtered, entry)
			}
		}
		entries = filtered
	}

	sort.Slice(entries, func(i, j int) bool {
		return entries[i].Date < entries[j].Date
	})

	return entries
}

// ──────────────────────────────────────────────────────────────
// String representation
// ──────────────────────────────────────────────────────────────

// String returns a human-readable representation.
// Implements the fmt.Stringer interface.
func (e *Exchange) String() string {
	return fmt.Sprintf("%s (%s)", e.Name, e.Code)
}